### PR TITLE
Caffe to Keras conversion

### PR DESCRIPTION
**Pull request by MarcBS**
This pull requests is a modified version of the one created by @pranv in #921. It adds the following improvements:
1. Fixed conversion of Caffe layers V2.
2. Fixed conversion to Graph models.
3. Added support for LRN (Local Response Normalization) and AVE (Average Pooling) layers.
4. Added command line conversion tool caffe2keras.py

Regards,
Marc

### CODE REVIEW

**Comment by fchollet** [comment]
Hi Marc,

At this stage I believe it would be preferable to create a separate repository for Caffe conversion, which will allow to provide more extensive tooling (command line utilities, etc).

LRN2D should be reintroduced as part of Keras itself, of course. However we need to get rid of the license dependency, which is the reason why I had removed it in the first place. Concretely that means rewriting the LRN2D layer from scratch.

**Comment by pranv** [comment]
This is really good work @MarcBS!

Regarding tests, I think it would be best if a test included importing of caffe python interface and performing `np.allclose(..)` to confirm caffe output and keras output are identical for a couple of models like GoogLeNet and VGG. 

Regrading moving to a separate repo/project, I think it's a good idea. If we can make it a bit more general and modular, it'll be useful in a lot of contexts outside keras (other theano/python libraries etc.,). We could have proper documentation and other utility functions around this to make a complete solution.

**Comment by pranv** [comment]
Something like this: https://github.com/NervanaSystems/caffe2neon/blob/master/tests/googlenet/test_googlenet.py

**Comment by MarcBS** [comment]
Thanks for your comments @fchollet and @pranv.

I have a fork ready with the modifications right here:
https://github.com/MarcBS/keras

I have also added a demo test file for applying a converted model on an image. Although I do not plan to invest a long time on this fork, I just needed the caffe2keras converter for my project, so if anybody wants to add more functionalities all of you are free to contribute to it! 

François, about your comment of the license dependency, does this mean that we can not include the LRN2D code in the fork either? In that case, if anybody wants to rewrite the layer feel free to submit a PR.

**Comment by pranv** [comment]
I'll look into making a standalone project in the following weeks then.

Anyone else interested in helping out?

**Comment by zo7** [comment]
What's the status of this? Did a separate project/repo ever get made?

**Comment by MarcBS** [comment]
I still have available a fork with the converter which I used for a project right here:
https://github.com/MarcBS/keras

I keep updating the fork with the new commits of the main repo regularly, although I am not planning to work actively on it.

**Comment by bkj** [comment]
Does anyone have examples of running the Keras and Caffe versions of the models side-by-side to show that they get the same results?

I tried to do this this evening and got different results from the two models, using the `keras2caffe` converter from https://github.com/hasnainv/keras and the `bvlc_reference_caffenet` model:

```
np.random.seed(123)
x = np.random.choice(range(255), (1, 3, 227, 227))

# Keras
go = K.function(
    [model.inputs[i].input for i in model.input_order],
    [model.nodes['norm1'].get_output(train=False)]
)

keras_result = go([x])[0]

# Caffe
net.blobs['data'].reshape(1, 3, 227, 227)
net.blobs['data'].data[0] = x
_ = net.forward()

caffe_result = net.blobs['norm1'].data

```

`keras_result` and `caffe_result` are _not_ the same -- the earlier layers `conv1` and `pool1` are the same, but `norm1` and beyond diverge .  Anyone have any thoughts on why the results might be different?

Thanks!
Ben

**Comment by Vishruit** [comment]
Hi Marc, 
This is really good. I am using some deep architectures for a project and this is a great piece of work for integrating the Caffe models to keras and it shall make it even easier to make run theano with the caffe models as the direct input. 
We tried the Caffe to Keras convertor, seems it works only for the v0.3 keras, if I am not wrong. As a lot of features have been added in keras, we shall like to help things ramp up and make it ready.
Thanks

**Comment by isaacgerg** [comment]
I would like this feature.  Is there a plan to roll this into Keras?

**Comment by MarcBS** [comment]
Hi @isaacgerg, thank you for your interest, but I think that there are no plans right now for introducing this feature to the main Keras repo. 

The main reasons are that @fchollet specifically asked to keep it in a separate fork due to some layers constraints, and also because I do not have time to regularly maintain and improve the converter, which in some cases would need to be updated for some newer layers and parameters introduced either in Keras or Caffe (PR are highly welcome :D!). 

I am working on adding several new layers to [my fork](https://github.com/MarcBS/keras), which can be useful for different multimodal learning problems I am currently working in. So, in case you were interested in using either these new layers or the caffe2keras converter you can use my fork (which also includes periodical updates to the main repo).

**Comment by anjishnu** [comment]
Hey Marc, any particular reason you aren't introducing a subset of your features to Keras? I'd like to play around with the CBP layer, for instance.

**Comment by MarcBS** [comment]
Hi @anjishnu, it is simply because of the lack of time to apply all the necessary tests on the different versions of Python and Theano/TF that the main Keras repo requires. Although, if you are using Python 2.7 and Theano you can directly use my fork, which also includes all the features in the main Keras repo.

**Comment by anjishnu** [comment]
Thanks Marc, I think the fork should be fine for now, once I get it working I'll see if we can contribute it. 

Best,
Anjishnu
